### ISSUE TITLE
Issues with duplicate OCIDs in record packages

### ISSUE THREAD

**Issue by jpmckinney** [open-contracting/kingfisher-process#420]
Context: It’s not possible to have a transaction run across both RabbitMQ and PostgreSQL. As such, it’s possible for e.g. RabbitMQ to deliver a message, Kingfisher Process to store the result in PostgreSQL, but for RabbitMQ to be shut down before Kingfisher Process can acknowledge the message to RabbitMQ. RabbitMQ will therefore re-deliver the message. Following best practice for message queues, Kingfisher Process can re-process duplicate messages, without introducing errors or interrupting processing.

With Chile Obras, multiple record packages contain the same OCID (like archivo_214244.json and archivo_283434.json for ocds-70d2nz-603128-10-LP13). `file_worker` just logs an integrity error, as above, which is then only logged to Sentry. Another possible case for duplication is if a single record package contains the same ocid multiple times; I don't think we've seen this.

I think **this causes the entire file containing the duplicate to be skipped**, since `_store_data()` iterates over each release/record, but the try/except is at the level of the message. Obviously, we'd like to continue processing the file, in case it contains some unique data.

This only occurs for record packages, because only the `record` table has a unique constraint on (collection, ocid) (the `release` table has no index on (collection, ocid, release_id) since we know publishers fail to maintain uniqueness like https://github.com/open-contracting/kingfisher-collect/issues/1049)

Options:

1. Remove the constraint. (I'd have to double-check if that affects anything else – mainly we'd be removing an index that is used in queries, but that can be restored via https://github.com/open-contracting/kingfisher-process/issues/329).
2. Add a try/except in `_store_data()`, as we don't expect duplicates at this level, and create a note if there's a duplicate.

In option (1), it'd be possible to check for duplicates by querying the database. (It would also be possible to see if the data for the duplicate OCIDs is identical or not.)

In terms of messages being idempotent, at present, it looks like duplicate messages for release packages are not idempotent, and new releases are saved (**we should add a test**). Duplicate releases have no impact on data processing, but if we're reporting such issues to publishers, we don't want such duplicates.

Duplicate messages for record packages are currently idempotent as discussed above.

If we want to make both idempotent: It's possible to disambiguate whether a message is the same or different, by querying the DB to see if the collection file for the release/record to be inserted is the same (meaning it’s a duplicate message) or different (meaning it’s a data issue) as the collection file for the release/record that's already in the DB.

**Comment by yolile** [comment]
> Obviously, we'd like to continue processing the file, in case it contains some unique data.

FWIW: which is the case in Chile

**Comment by jpmckinney** [comment]
Which Chile spider is best to test with?

**Comment by yolile** [comment]
> Which Chile spider is best to test with?

We don't have a spider for it yet, only a local load. The data is available at `/home/ylisnichuk/local-load/chile-obras-publicas-2024-03-20`. I will retry it now.

**Comment by jpmckinney** [comment]
I haven’t progressed this issue yet

**Comment by jpmckinney** [comment]
Would this also affect https://github.com/open-contracting/kingfisher-collect/issues/965 ? Where panama duplicated a record at the start of the next page (in which case, I think the remaining records on that page wouldn't get imported)

**Comment by yolile** [comment]
> Would this also affect https://github.com/open-contracting/kingfisher-collect/issues/965 ?

Yes, I think so

**Comment by jpmckinney** [comment]
> Remove the constraint. (I'd have to double-check if that affects anything else)

I added the constraint in e6eaf6a11451478d1d2895521623b6de490a791b, because the record_compiler worker will raise Record.MultipleObjectsReturned if there are multiple records with the same OCID. This makes sense – there is no way to choose which record to compile. I added that on Sep 23, 2022, same day that `panama_dgcp_records` was added to Collect – so, maybe related.

If we remove the constraint, we can add code to that worker to pick the first one (e.g. order by `id`), and if there are more, then log an error.
